res/controllers/Vestax-VCI-100MKII-scripts.js

```diff
@@ -158,164 +208,165 @@ VCI102.pitch = function(ch, midino, value, status, group) {
+    var n = engine.getValue(group, "focused_effect");
+    if (n) {
+        VCI102.parameter(ch, VCI102.slot(group, n), "parameter3", value);
+    } else {
+        VCI102.set(VCI102.slot(group, 3), "meta", value);
+    }
 };
 
-VCI102.loop = function(ch, midino, value, status, group) {
+VCI102.mix_enabled = [0, 0, 0, 0];
+VCI102.value4 = [0, 0, 0, 0];  // recent value of parameter4 knob
+
+VCI102.mix_toggle = function(ch, midino, value, status, group) {
     if (value) {
-        if (engine.getValue(group, "loop_enabled")) {
-            engine.setValue(group, "reloop_exit", 1);
-        } else {
-            engine.setValue(group, "beatloop", VCI102.loopLength[ch]);
+        VCI102.mix_enabled[ch] ^= 127;
+        if (VCI102.mix_enabled[ch]) {
+            engine.softTakeoverIgnoreNextValue(group, "mix");
+            VCI102.set(group, "mix", VCI102.value4[ch]);  // prevent slip
         }
+        midi.sendShortMsg(status, midino, VCI102.mix_enabled[ch]);
     }
 };
 
-VCI102.reloop = function(ch, midino, value, status, group) {
-    if (engine.getValue(group, "loop_enabled")) {
-        engine.setValue(group, "loop_out", value > 0);
+VCI102.parameter4 = function(ch, midino, value, status, group) {
+    var n = engine.getValue(group, "focused_effect");
+    if (VCI102.mix_enabled[ch]) {
+        VCI102.set(group, "mix", value);
+    } else if (n) {
+        VCI102.parameter(ch, VCI102.slot(group, n), "parameter4", value);
     } else {
-        engine.setValue(group, "reloop_exit", 1);
+        VCI102.super1(ch, midino, value, status, group);
     }
+    VCI102.value4[ch] = value;
 };
 
-VCI102.loop_halve = function(ch, midino, value, status, group) {
-    if (engine.getValue(group, "loop_enabled")) {
-        engine.setValue(group, "loop_halve", value > 0);
-    } else if (value) {
-        VCI102.setLoopLength(ch, status, VCI102.loopLength[ch] / 2);
+VCI102.no_slot = function(ch, midino, value, status, group) {
+    if (value) {
+        engine.setValue(group, "focused_effect", 0);
     }
 };
 
-VCI102.loop_double = function(ch, midino, value, status, group) {
-    if (engine.getValue(group, "loop_enabled")) {
-        engine.setValue(group, "loop_double", value > 0);
-    } else if (value) {
-        VCI102.setLoopLength(ch, status, VCI102.loopLength[ch] * 2);
+VCI102.next_slot = function(ch, midino, value, status, group) {
+    if (value) {
+        engine.setValue(group, "focused_effect",
+                        engine.getValue(group, "focused_effect") % 3 + 1);
+    }
+};
+
+VCI102.loop = function(ch, midino, value, status, group) {
+    if (value) {
+        if (engine.getValue(group, "loop_enabled")) {
+            engine.setValue(group, "reloop_toggle", 1);
+        } else {
+            engine.setValue(group, "beatloop_activate", 1);
+            engine.setValue(group, "beatloop_activate", 0);
+        }
+    }
+};
+
+VCI102.reloop = function(ch, midino, value, status, group) {
```

Comment:
I got quite confused figuring out how this is working. Could you add a comment explaining that this is the `loop` button with shift?

Reply:
Some old controllers including VCI-100mkII send messages with different midi numbers depending on shift. So not only VCI102.loop(without shift)-VCI102.reloop(with shift), but also other controls, play-reverse, sync_enabled-quantize, ... can be mapped to same buttons without scripts. You can see them in "Vestax VCI-100MKII.midi.xml".